### ISSUE TITLE
missed opportunities to introduce and/or

### ISSUE THREAD

**Issue by jvoigtlaender** [ndmitchell/hlint#1049]
The hints https://github.com/ndmitchell/hlint/blob/0f41abcd32f8b024c25f138aab99c23e69bc5270/data/hlint.yaml#L205-L206
work nicely on code like `x` in the following, but not on code like `y`:
```haskell
x = elem True []

y = True `elem` []
```
That is, the following is reported by `hlint` on these definitions:
```
test.hs:1:5-13: Warning: Use or
Found:
  elem True
Perhaps:
  or

test.hs:1:5-16: Suggestion: Use infix
Found:
  elem True []
Perhaps:
  True `elem` []

2 hints
```
Both hints refer to `x`.

One could have hoped that there would also be a hint reported on `y`, along the lines of:
```
Found:
  True `elem` []
Perhaps:
  or []
```
But appatently, the hint
```yaml
    - warn: {lhs: elem True, rhs: or}
```
does not apply to infix uses of `elem`. I don't know enough about the rule matching process in `hlint` to understand whether it should.

In any case, it would of course be possible to add the following two extra rules:
```yaml
- warn: {lhs: True `elem` l, rhs: or l}
- warn: {lhs: False `notElem` l, rhs: and l}
````
But it's unclear to me whether that's the way to go.

**Comment by ndmitchell** [comment]
I just tried this on HLint HEAD:

```
y = True `elem` []
```

Gives:

```
Sample.hs:1:5-18: Warning: Use or
Found:
  True `elem` xs
Perhaps:
  or xs
```

Which HLint are you using? I seem to remember fixing this bug at some point in the past, but not 100% sure, and can't find it in the changelog. First thing to do is synchronise on why we are seeing different things.

**Comment by jvoigtlaender** [comment]
My observation was with hlint-3.1.4.

**Comment by ndmitchell** [comment]
I just switched to hlint 3.1.4 and used the file Sample.hs containing:

```
y = True `elem` []
```

It warned. Can you try a file with just that in it?

**Comment by jvoigtlaender** [comment]
Hmm, strange. At my end:
```
> cat .\sample.hs
y = True `elem` []
> hlint --version
HLint v3.1.4, (C) Neil Mitchell 2006-2020
> hlint .\sample.hs
No hints
```

**Comment by jvoigtlaender** [comment]
That's on Windows. Installed with `cabal install hlint` a few days ago.

**Comment by ndmitchell** [comment]
I tried Windows, a fresh install, and I got no hints - so the issue must be at my end. I'll take another look.

**Comment by ndmitchell** [comment]
I don't think it's unreasonable that the matching engine _could_ make it match, but it requires the matching engine remembering there is some stuff that needs adding back at the end of the match. We do that for prefix stuff, but not yet for suffix stuff - and the prefix stuff is a bit of a nightmare, so given there are only two rules we have where the values are likely to be infix with backticks, I've gone for the simpler option of adding specific infix versions, along the lines you wrote in the initial comment.

**Comment by jvoigtlaender** [comment]
Makes perfect sense!
